File: Binary Search/Hard/LET_no668_Kth Smallest Number in Multiplication Table.py
class Solution:
    def findKthNumber(self, m: int, n: int, k: int) -> int:
        def enough(num):
            return sum(min(num // i, n) for i in range(1, m + 1)) >= k
        
        low, high = 1, m * n
        while low + 1 < high:
            mid = (low + high) // 2
            if not enough(mid):
                low = mid
            else:
                high = mid
        
        if enough(low):
            return low
        return high

File: Binary Search/Hard/LET_no668_Kth Smallest Number in Multiplication Table.cpp
class Solution {
public:
    int findKthNumber(int m, int n, int k) {
        int low = 1, high = m * n;
        while (low + 1 < high) {
            int mid = (low + high) / 2;
            if (!enough(mid, m, n, k))
                low = mid;
            else
                high = mid;
        }
        
        if (enough(low, m, n, k))
            return low;
        return high;
    }
    
    bool enough(int num, int m, int n, int k) {
        int res = 0;
        for (int i = 1; i <= m; i++) {
            res += min(num / i, n);
        }
        return res >= k;
    }
};